Code for adding more options for parameter space and frequency space interpolation for SEOBNRv2-ROMs

soumide1102 opened the pull request:
* @cdcapano @duncan-brown I've added the new algorithm  ```partial_compress_rom``` for doing the parameter space interpolation in the SEOBNRv2_ROM LAL code and importing the amplitude and phase interpolants from there. The function will then call ```fd_decompress``` to do the frequency space interpolation. This function would require a modification in the LAL code too so that it allows an access to the amplitude and frequency interpolants directly. See https://bugs.ligo.org/redmine/issues/4825 . 

* I've added a new command line argument in ```pycbc_compress_bank``` called ```--ROM-parameter-space-interpolation``` which would which would take in options ```linear_compression``` or ```partial_ROM_compression``` to decide the process of ROM compression.

* I've added modifications in ```compress.py``` so that it supports handling and storing two separate set of frequency points for amplitude and phase when doing the new partial ROM compression and deals with one set of frequency points for the spa compression. This takes into consideration the changes requested in https://github.com/ligo-cbc/pycbc/pull/1185 .

@cdcapano could you please take a look at this?

duncan-brown commented:
@soumide1102 there are still landscape issues. Pleae can you fix? Please remove the work in progress label when this is ready to look at.

soumide1102 commented:
@duncan-brown  @cdcapano I've fixed the landscape issues. Could you please take a look?

johnveitch commented:
https://bugs.ligo.org/redmine/issues/4825 was pushed to LALSimulation

soumide1102 commented:
Thanks @johnveitch !

duncan-brown commented:
Hi John,

Great, thank you!

Cheers,
Duncan.

> On Nov 11, 2016, at 7:49 AM, John Veitch <notifications@github.com> wrote:
> 
> https://bugs.ligo.org/redmine/issues/4825 was pushed to LALSimulation
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 

-- 

Duncan Brown                         http://dbrown10.expressions.syr.edu
Charles Brightman Professor of Physics     Room 263-1 Physics Department
Director of the Graduate Program      Syracuse University, NY 13244, USA
Phone: 315 443 5993                                    Fax: 315 443 9103

duncan-brown commented:
@cdcapano @josh-willis this branch is ready to merge. Can we go ahead?

cdcapano commented on bin/pycbc_compress_bank:
```diff
@@ -146,23 +169,57 @@ for ii in range(imin, imax):
+
+    # If template is SEOBNRv2_ROM perform linear_compression or
+    # partial_ROM_compression as per user input
+    if mtotal >= 4:
```
This is bad... choices about what parameters an approximant is used for should not be hard coded. This makes this code only useable on the O1 template bank.

soumide1102 commented:
Hi @cdcapano . Thanks for your comments. I'll make the changes you requested.

soumide1102 commented:
@cdcapano I've added in the changes you had requested. 
The major changes are as follows :
1. `pycbc_compress_bank` would be able to take in multiple arguments for `--compression-algorithm` so that separate algorithms can be specified for TaylorF2 and SEOBNRv2_ROMs or a common algorithm which would work for both can be entered. I've changed the compression-algorithm keys in compress.py accordingly. 
2. I've changed the calculation of `mismatch` in `partial_rom_compression` from `1.0 - overlap` to `1.0 - match` as calculating the match would be more appropriate in this case. 

Could you please take a look at the updated code?

duncan-brown commented:
@cdcapano bump on this

@soumide1102 please rebase the branch to master as the pull request is pretty old now.

soumide1102 commented:
Rebased. 

Landscape showed some issues after rebasing about which it had never complained before. So I've fixed them.

duncan-brown commented:
@cdcapano I'm assuming silence is no objection, so I will merge this later today.

cdcapano commented on bin/pycbc_compress_bank:
```diff
@@ -40,72 +40,105 @@ from pycbc import filter
+                    "algorithm compatible with both. For example"
+                    "partial_rom_compression works only for the ROMs."
+                    "Look at compress.compression_algorithms.keys()"
+                    "for the available options")
```
Don't make the combination of approximant and algorithms possible choices. Instead, just list here the possible algorithms, and make it so that an approximant name can be added to only apply that algorithm to templates with that approximant. Also don't limit the number of algorithms that may be used to 2. If at some point we end up using more than two approximants in a bank (as we nearly did for O2), you'll want the code to be able to support that. I suggest something like:
```
parser.add_argument("--compression-algorithm", nargs='+', type=str,
                    required=True, metavar="ALGORITHM[:APPRX]",
                    help="Specify the compression algorithm to use. Choices "
                         "are {}. ".format(
                         ', '.join(compress.compression_algorithms.keys())) +
                         "To only use an algorithm with a specific "
                         "approximant, add the approximant name preceeded by "
                         "a colon; e.g., 'spa:TaylorF2'. In this case, an "
                         "algorithm must be provided for every approximant "
                         "that exists in the bank.")
```

cdcapano commented on bin/pycbc_compress_bank:
```diff
@@ -40,72 +40,105 @@ from pycbc import filter
+    compression_type_approximant1, compression_type_approximant2 = args.compression_algorithm
+    compression_type1, approximant1 = compression_type_approximant1.split(':')
+    compression_type2, approximant2 = compression_type_approximant2.split(':')
+
```
As I mentioned above, don't hard code the number of arguments to be 1 or 2. You can deal with the arbitrary case by creating a dictionary mapping approximant name -> compression algorithm. To ensure that the command encompasses all of the approximants in the bank, move this parsing to after the bank is loaded, so you can get all the approximants present in the bank. Then, when done parsing the `compression_alrgorithm` option, check that all of the approximants are covered, raising an error otherwise.

cdcapano commented on bin/pycbc_compress_bank:
```diff
@@ -146,23 +178,40 @@ for ii in range(imin, imax):
+        if tmplt_approximant == approximant1:
+            compression_type = compression_type1
+        if tmplt_approximant == approximant2:
+            compression_type = compression_type2
```
Having created the approximant name -> compression algorithm dictionary above, here you could get the `compression_type` by just passing this template's approximant to the dictionary.

cdcapano commented on bin/pycbc_compress_bank:
```diff
@@ -146,23 +178,40 @@ for ii in range(imin, imax):
+    elif compression_type == 'partial_compress_rom':
+        hcompressed = compress.partial_rom_compression(
+            htilde, tmplt.mass1, tmplt.mass2, tmplt.spin1z, tmplt.spin2z,
+            df, fmin, (kmax-1)*df, args.interpolation)
```
It might be good to move this decision making about what compress function to call as a function to the compress module (à la `get_fd_waveform`). Then, as you add compression algorithms in the future, you need only edit the compress module and not touch the executable. But I'm fine with leaving this here for now; that can be addressed in a future PR.

cdcapano commented on pycbc/waveform/compress.py:
```diff
@@ -260,24 +258,148 @@ def compress_waveform(htilde, sample_points, tolerance, interpolation,
+        'spa:SEOBNRv2_ROM_DoubleSpin' : spa_compression,
+        'mchirp:TaylorF2' : mchirp_compression,
+        'mchirp:SEOBNRv2_ROM_DoubleSpin' : mchirp_compression
+        }
```
The compression algorithms should only list the compression algorithm names, not the approximants.

soumide1102 commented:
@cdcapano Thanks for the comments! I had thought of adding the compression algorithm options for SEOBNRv4_ROM after we get the partial compression algorithm working for it. So I kept the approximants limited to TaylorF2 and SEOBNRv2_ROM_DoubleSpin. However, I agree that it would be nice to give the code the ability to read in more than two kinds of approximants from a bank. I'll make the changes you mentioned and update the PR.

soumide1102 commented:
@cdcapano I have updated the PR with the changes you had requested. `--compression-algorithm` would now be able to read in multiple arguments to deal with more than two approximants in the bank. Also, `compression_algorithms` in `compress.py` now lists the possible algorithms and *not* combinations of the algorithms and possible approximants. Could you please take a look?

spxiwh commented:
Any updates on this? If there's nothing new the next time I look over old pull requests then I'll close this ... It's easy to open a new pull request later.

spxiwh commented:
Closing as this has seen no activity in almost a year.
